File: backend/src/models/TeacherProfile.model.js
import mongoose from 'mongoose';

const teacherProfileSchema = new mongoose.Schema({

    userId: {
        type: mongoose.Schema.Types.ObjectId, 
        ref: 'User',
        required: true,
        unique: true,
    },

    teacherRegId: {  // from this teacher can login
        type: String,
        required: true,
        unique: true,
        trim: true,
        uppercase: true
    },

    cnic: {
        type: String,
        required: true,
        unique: true,
        match: [/^\d{5}-\d{7}-\d$/, 'CNIC format: 42123-1234567-1'],
    },

    qualification: {
        type: String,
        required: true,
    },

    salary: {
        type: Number,
        required: true,
        min: 0,
    },

    joiningDate: {
        type: Date,
        default: Date.now,
    },

    subjects: [String],

    classes: {  // new field added
        type: Number,
        required: true
    },  
    
    contact: {
        type: String,
        required: true
    },
    
    address: {
        type: String,
        required: true
    },

    age: {
        type: Number,
        required: true
    },
}, { timestamps: true });

// teacherProfileSchema.index({ cnic: 1 }); // CNIC ke basis pr fast searching ke liye

export default mongoose.model('TeacherProfile', teacherProfileSchema);

File: MobileApp/irp_mobile_app/lib/models/teacher_profile_model.dart
class TeacherProfileModel {
  final String id;
  final String userId;
  final String teacherFullName;
  final String teacherRegId;
  final String cnic;
  final String qualification;
  final double salary;
  final DateTime joiningDate;
  final List<String> subjects;
  final List<int> classes;
  final String contact;
  final String address;
  final int age;

  TeacherProfileModel({
    required this.id,
    required this.userId,
    required this.teacherFullName,
    required this.teacherRegId,
    required this.cnic,
    required this.qualification,
    required this.salary,
    required this.joiningDate,
    required this.subjects,
    required this.classes,
    required this.contact,
    required this.address,
    required this.age,
  });

  factory TeacherProfileModel.fromJson(Map<String, dynamic> json) {
    return TeacherProfileModel(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? '',
      teacherFullName: json['teacherFullName'] ?? '',
      teacherRegId: json['teacherRegId'] ?? '',
      cnic: json['cnic'] ?? '',
      qualification: json['qualification'] ?? '',
      salary: (json['salary'] ?? 0).toDouble(),
      joiningDate: json['joiningDate'] != null
          ? DateTime.parse(json['joiningDate'])
          : DateTime.now(),
      subjects: List<String>.from(json['subjects'] ?? []),
      classes: List<int>.from(json['classes'] ?? []),
      contact: json['contact'] ?? '',
      address: json['address'] ?? '',
      age: json['age'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'userId': userId,
      'teacherFullName': teacherFullName,
      'teacherRegId': teacherRegId,
      'cnic': cnic,
      'qualification': qualification,
      'salary': salary,
      'joiningDate': joiningDate.toIso8601String(),
      'subjects': subjects,
      'classes': classes,
      'contact': contact,
      'address': address,
      'age': age,
    };
  }
}
